src/librbd/io/ImageRequestWQ.cc

```diff
@@ -452,6 +452,20 @@ void ImageRequestWQ::process(ImageRequest<> *req) {
   ldout(cct, 20) << "ictx=" << &m_image_ctx << ", "
                  << "req=" << req << dendl;
 
+  if (m_image_ctx.get_max_iops()) { //if we set limit value
+    utime_t curr_L = m_image_ctx.qos_status->get_prev_limit();
+    utime_t now = ceph_clock_now();
+
+    curr_L += (double) (1.0 / m_image_ctx.get_max_iops());
+    curr_L = (curr_L > now ? curr_L : now);
+    if (curr_L > now) {
+      ldout(cct, 20) << "we are ahead by " << curr_L - now << ", sleep" << dendl;
+      utime_t delay = curr_L - now;
+      delay.sleep();
```

Comment:
That's not ideal either since you will block whatever thread is pushing IO into librbd.